### PR TITLE
Refactor sparse masks + OFT

### PR DESCRIPTION
**Pull request by oleksost**
- refactored sparse mask
- moved mask update functionality into SPLITExpertModule. In this way we (1) can trigger mask updates after the current batch is used (so no batch is wasted) and (2) we do not run the risk of messing up the optimizer step as we do not call .step() when updating the mask (only backward)
- make sure can repro Samin's results
(compared on wiqa_what_is_the_final_step_of_the_following_process)
image

### CODE REVIEW

File: mttl/models/modifiers/oft.py [hunk 1]
```diff
@@ -0,0 +1,605 @@
...
+            else None
+        )
+        result = F.linear(input=x, weight=scaled_rotated_weight, bias=bias)
+
+        result = result.to(previous_dtype)
+        return result
+        
+
+
+# class Linear(nn.Module, OFTLayer):
```
Comment (sordonia) [inline mttl/models/modifiers/oft.py:410]: remove this comment?

File: mttl/models/modifiers/sparse_mask.py [hunk 2]
```diff
@@ -26,637 +30,53 @@
...
     )
-    reselection_steps: int = 100  # every how many steps to switch to mask update regime
-    sps_type: str = "block_sparse"  # ['block_sparse','regular_sparse','row_sparse']
-    sps_impl: str = "sp_add+sp_mm"  # ['sp_add+sp_mm','scattered', 'masked_linear']
-    selection_algorithm: str = "rigl"
-    reselection_rate_policy: str = "linear"
+    mask_reselection_interval: int = (
+        100  # every how many steps to switch to mask update regime
+    )
+    n_max_mask_reseleciton: int = (
```
Comment (sordonia) [inline mttl/models/modifiers/sparse_mask.py:40]:
```suggestion
    n_max_mask_reselection: int = (
```

File: mttl/models/modifiers/sparse_mask.py [hunk 3]
```diff
@@ -26,637 +30,53 @@
...
-@MaskUpdatWrapper.register("snip", config_cls=SparseMaskConfig)
-class SNIPMaskUpdateWrapper(MaskUpdatWrapper):
-    """
-    SNIPMaskUpdateWrapper is a wrapper around SparseLinear.
+    SNIPMaskUpdater is a wrapper around SparseLinear.
     It is used to periodically re-calculate the sparse mask indices a la SNIP (https://arxiv.org/pdf/1810.02340).
     To recalculate the mask, it uses a couple of incoming mini-batches to estimate the importance of each parameter.
 
-    This method willl also keep track of the selected weights throughout learning on CPU.
+    If accumulates learned weights in a dense CPU matrix. This is useful e.g. to make sure that the weights that have been learned in the past and are selected again are not reinitialized to 0.
```
Comment (sordonia) [inline mttl/models/modifiers/sparse_mask.py:65]:
```suggestion
    It accumulates learned weights in a dense CPU matrix. This is useful e.g. to make sure that the weights that have been learned in the past and are selected again are not reinitialized to 0.
```

File: mttl/models/modifiers/sparse_mask.py [hunk 4]
```diff
@@ -748,47 +183,52 @@ def selected_params(self) -> torch.Tensor:
...
+    def _time_to_update_sparse_weights(self, sparse_layer: SparseLinear):
         return (
-            self._mask_update_steps % self.config.selection_accumulation_steps == 0
-            and self.sparse_layer.training
+            self._mask_update_steps % self.config.steps_in_mask_selection == 0
+            and sparse_layer.training
         )
 
-    def prepate_mask_or_weights_learning(self):
+    def prepate_mask_or_weights_learning(self, sparse_layer: SparseLinear):
```
Comment (sordonia) [inline mttl/models/modifiers/sparse_mask.py:203]:
```suggestion
    def prepare_mask_or_weights_learning(self, sparse_layer: SparseLinear):
```

File: mttl/models/modifiers/sparse_mask.py [hunk 5]
```diff
@@ -717,24 +143,33 @@ def mask_backward_hook(mask):
         )
         self._backward_hooks.append(hook_handle)
 
-    def switch_to_weights_update_modus(self):
+    def switch_to_weights_update_modus(self, sparse_layer: SparseLinear):
```
Comment (sordonia) [inline mttl/models/modifiers/sparse_mask.py:146]:
```suggestion
    def switch_to_weights_update_mode(self, sparse_layer: SparseLinear):
```

**Review by sordonia — APPROVED** [review]

File: mttl/models/modifiers/oft.py (hunk 1 above)
Comment (oleksost) [reply mttl/models/modifiers/oft.py:410]: addressed

File: mttl/models/lightning/expert_module.py [hunk 6]
```diff
@@ -144,6 +149,50 @@ def on_save_checkpoint(self, ckpt):
         ).expert_info.asdict()
 
 
+class SPLITExpertModule(ExpertModule):
+    """
+    Expert module used to train sparse mask with SPLIT mask updater.
```
Comment (pclucas14) [inline mttl/models/lightning/expert_module.py:154]: what is SPLIT mask updater ? Can we add a paper reference ?

File: mttl/models/lightning/expert_module.py [hunk 7]
```diff
@@ -144,6 +149,50 @@ def on_save_checkpoint(self, ckpt):
         ).expert_info.asdict()
 
 
+class SPLITExpertModule(ExpertModule):
```
Comment (pclucas14) [inline mttl/models/lightning/expert_module.py:152]: Consider putting in a separate file

File: mttl/models/modifiers/oft.py [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+                f"Either `r` or `oft_block_size` must be non-zero. Currently, r = {self.r} and oft_block_size = {self.oft_block_size}."
+            )
+        if not (self.r != 0) ^ (self.oft_block_size != 0):
+            raise ValueError(
+                f"You can only specify either r ({self.r}) or oft_block_size ({self.oft_block_size}), but not both simultaneously, because r x oft_block_size == in_features."
+            )
+
+
+@Modifier.register("oft", config_cls=OFTConfig)
+class OFTLayer(Modifier, ModifyMixin):
```
Comment (pclucas14) [inline mttl/models/modifiers/oft.py:88]: Add a ref to the OFT paper ?

File: projects/modular_llm/train_experts.py [hunk 9]
```diff
@@ -203,7 +203,7 @@ def upload_library(expert_library, module):
                             expert = module.get_expert_instance(expert_name)
                             expert_library.add_expert(expert, expert_name)
                 elif isinstance(module, ExpertModule):
-                    expert = module.as_expert()
+                    expert = module.model.as_expert()
```
Comment (pclucas14) [inline projects/modular_llm/train_experts.py:206]: What changed here for this to be required ?

**Review by pclucas14 — APPROVED** [review]
Sparse Masks FTW

File: mttl/models/lightning/expert_module.py (hunk 6 above)
Comment (oleksost) [reply mttl/models/lightning/expert_module.py:154]: addressed

File: mttl/models/modifiers/oft.py (hunk 8 above)
Comment (oleksost) [reply mttl/models/modifiers/oft.py:88]: addressed

File: mttl/models/lightning/expert_module.py (hunk 7 above)
Comment (oleksost) [reply mttl/models/lightning/expert_module.py:152]: addressed

File: mttl/models/lightning/expert_module.py (hunk 6 above)
Comment (oleksost) [reply mttl/models/lightning/expert_module.py:154]: addressed

File: projects/modular_llm/train_experts.py (hunk 9 above)
Comment (oleksost) [reply projects/modular_llm/train_experts.py]: addressed

**Review by sordonia — APPROVED** [review]
